packages/@lwc/engine/src/framework/template.ts

```diff
@@ -151,17 +149,19 @@ export function evaluateTemplate(vm: VM, html: Template): Array<VNode | null> {
             // job
             const { component, context, cmpSlots, cmpTemplate, tro } = vm;
             tro.observe(() => {
-                // reset the cache memoizer for template when needed
+                // Reset the cache memoizer for template when needed.
                 if (html !== cmpTemplate) {
-                    // perf opt: do not reset the shadow root during the first rendering (there is nothing to reset)
-                    if (!isUndefined(cmpTemplate)) {
-                        // It is important to reset the content to avoid reusing similar elements generated from a different
-                        // template, because they could have similar IDs, and snabbdom just rely on the IDs.
+                    // Perf opt: do not reset the shadow root during the first rendering (there is
+                    // nothing to reset).
+                    if (!isNull(cmpTemplate)) {
+                        // It is important to reset the content to avoid reusing similar elements
+                        // generated from a different template, because they could have similar IDs,
+                        // and snabbdom just rely on the IDs.
                         resetShadowRoot(vm);
                     }
 
-                    // Check that the template was built by the compiler
-                    if (isUndefined(html) || !isTemplateRegistered(html)) {
+                    // Check that the template was built by the compiler.
+                    if (isNull(html) || !isTemplateRegistered(html)) {
```

Comment:
Why are we now checking if `html` it not null? It can also be undefined. We should either guard both the falsy values or let isTemplateRegistered() handle both.

Reply:
The passed `html` has to be a `Template`. I think we can remove the `isNull` check. Can you explain how `html` might be `undefined` here?

Reply:
In the following case, in PROD mode, `html` at this point will be `undefined`
```js
MyComponent extends LightningElement {
     render() {
          return; // For some reason, if the render returns an undefined.
     }
}
```

Reply:
Good catch, I removed the `isNull` check. While this doesn't prevent the case an `undefined` template reaches this part of the engine, it will throw an error because we should never register an `undefined` template.